includes/admin/class.llms.admin.setup.wizard.php

```diff
@@ -260,130 +320,206 @@ public function output() {
 		$next      = $this->get_next_step();
 
-		if ( in_array( $current, array_keys( $this->get_steps(), true ) ) ) {
+		if ( in_array( $current, array_keys( $this->get_steps() ), true ) ) {
 
 			ob_start();
 			include LLMS_PLUGIN_DIR . 'includes/admin/views/setup-wizard/step-' . $current . '.php';
 			$step_html = ob_get_clean();
 
 		}
 
+		/**
+		 * Filter the HTML of a step within the setup wizard.
+		 *
+		 * The dynamic portion of this hook, `$current`, refers to the slug of the current step.
+		 *
+		 * This filter can be used to output the HTML for a custom step in the setup wizard.
+		 *
+		 * @since [version]
+		 *
+		 * @param string                  $step_html HTML of the step.
+		 * @param LLMS_Admin_Setup_Wizard $wizard    Setup wizard class instance.
+		 */
+		$step_html = apply_filters( "llms_setup_wizard_{$current}_html", $step_html, $this );
+
 		include LLMS_PLUGIN_DIR . 'includes/admin/views/setup-wizard/main.php';
 
 	}
 
+	/**
+	 * Output HTML prior to each importable course
+	 *
+	 * Adds an opening label wrapper and adds HTML data to turn the element into a checkbox.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $course Importable course data array.
+	 * @return void
+	 */
+	public function output_before_importable_course( $course ) {
+		?>
+		<label>
+			<input name="llms_setup_course_import_ids[]" value="<?php echo absint( $course['id'] ); ?>" type="checkbox" checked>
+			<span class="dashicons dashicons-yes-alt"></span>
+		<?php
+	}
+
+	/**
+	 * Output HTML after to each importable course
+	 *
+	 * Closes the label element opened in `output_before_importable_course()`.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $course Importable course data array.
+	 * @return void
+	 */
+	public function output_after_importable_course( $course ) {
+		echo '</label>';
+	}
+
 	/**
 	 * Handle saving data during setup
 	 *
 	 * @since 3.0.0
 	 * @since 3.3.0 Unknown.
 	 * @since 3.35.0 Sanitize input data; load sample data from `sample-data` directory.
 	 * @since 3.37.14 Ensure redirect to proper course when a course is imported at the end of setup.
+	 * @since [version] Moved logic for each wizard step into it's own method.
 	 *
-	 * @return void
+	 * @return null|WP_Error
 	 */
 	public function save() {
 
-		if ( ! isset( $_POST['llms_setup_nonce'] ) || ! llms_verify_nonce( 'llms_setup_nonce', 'llms_setup_save' ) ) {
-			return;
+		if ( ! isset( $_POST['llms_setup_nonce'] ) || ! llms_verify_nonce( 'llms_setup_nonce', 'llms_setup_save' ) || ! current_user_can( 'manage_lifterlms' ) ) {
+			return null;
 		}
 
-		if ( ! current_user_can( 'manage_lifterlms' ) ) {
-			return;
+		$res = new WP_Error( 'llms-setup-save-invalid', __( 'There was an error saving your data, please try again.', 'lifterlms' ) );
+
+		$step = llms_filter_input( INPUT_POST, 'llms_setup_save', FILTER_SANITIZE_STRING );
+		if ( method_exists( $this, 'save_' . $step ) ) {
+			$res = call_user_func( array( $this, 'save_' . $step ) );
```

Comment:
I continue on my golf (and perf) theory :P 
```suggestion
			$res = $this->{"save_$step"}();
```
look at that beauty!

Reply:
I still haven't seen any metrics on your performance theory though.... wink

Reply:
a metric that certifies that function calls are expensive in php in general
or that despite we know they're expensive in php in general `call_user_func( array( $this, 'save_' . $step ) );` is still less expensive than `$this->{"save_$step"}();`? :D

Reply:
Still no comment on the beauty though...